crates/sui-adapter/src/programmable_transactions/execution.rs

```diff
@@ -436,6 +446,93 @@ fn execute_move_publish<E: fmt::Debug, S: StorageView<E>, Mode: ExecutionMode>(
     Ok(values)
 }
 
+/// Upgrade a Move package.  Returns an `UpgradeReceipt` for the upgraded package on success.
+fn execute_move_upgrade<E: fmt::Debug, S: StorageView<E>, Mode: ExecutionMode>(
+    context: &mut ExecutionContext<E, S>,
+    upgrade_ticket_arg: Argument,
+    dep_ids: Vec<ObjectID>,
+    module_bytes: Vec<Vec<u8>>,
+) -> Result<Vec<Value>, ExecutionError> {
+    // Check that package upgrades are supported.
+    context
+        .protocol_config
+        .check_package_upgrades_supported()
+        .map_err(|_| ExecutionError::from_kind(ExecutionErrorKind::FeatureNotYetSupported))?;
+
+    assert_invariant!(
+        !module_bytes.is_empty(),
+        "empty package is checked in transaction input checker"
+    );
+
+    let upgrade_ticket: UpgradeTicket = {
+        let mut ticket_bytes = Vec::new();
+        let ticket_val: Value =
+            context.by_value_arg(CommandKind::Upgrade, 0, upgrade_ticket_arg)?;
+        let ticket_type = context
+            .session
+            .load_type(&TypeTag::Struct(Box::new(UpgradeTicket::type_())))
+            .map_err(|e| context.convert_vm_error(e))?;
+        check_param_type::<_, _, Mode>(context, 0, &ticket_val, &ticket_type)?;
+        ticket_val.write_bcs_bytes(&mut ticket_bytes);
+        bcs::from_bytes(&ticket_bytes).map_err(|_| {
+            ExecutionError::from_kind(ExecutionErrorKind::CommandArgumentError {
+                arg_idx: 0,
+                kind: CommandArgumentError::InvalidBCSBytes,
+            })
+        })?
+    };
+
+    // Run the move + sui verifier on the modules and publish them into the cache.
+    let upgraded_package_modules = publish_and_verify_modules::<_, _, Mode>(context, module_bytes)?;
```

Comment:
We shouldn't do this until we have checked compatibility, otherwise we're going to pollute the loader unnecessarily.

Reply:
Sounds good. I originally had it lower down, but lifetime issues made me want to move it up here. I'll fix it :)

Reply:
Actually one question I had here is do we want to run the normalizer before we perform verification? From what I could tell poking around the code it seemed like a "this is fine, but you should really run the bytecode verifier first before doing this" which would mean we'd want to verify and publish first?